LoRaEngine/LoraKeysManagerFacade/EdgeDeviceGetter.cs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoraKeysManagerFacade
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using LoRaTools;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+
+    internal class EdgeDeviceGetter
+    {
+        private readonly IDeviceRegistryManager registryManager;
+        private readonly ILoRaDeviceCacheStore cacheStore;
+        private readonly ILogger<EdgeDeviceGetter> logger;
+
+        public EdgeDeviceGetter(IDeviceRegistryManager registryManager,
+                                ILoRaDeviceCacheStore cacheStore,
+                                ILogger<EdgeDeviceGetter> logger)
+        {
+            this.registryManager = registryManager;
+            this.cacheStore = cacheStore;
+            this.logger = logger;
+        }
+
+        private async Task<List<Twin>> GetEdgeDevicesAsync()
+        {
+            this.logger.LogDebug("Getting Azure IoT Edge devices");
+            var q = this.registryManager.CreateQuery("SELECT * FROM devices where capabilities.iotEdge = true");
+            var twins = new List<Twin>();
+            do
+            {
+                twins.AddRange(await q.GetNextAsTwinAsync());
+            } while (q.HasMoreResults);
+            return twins;
+        }
+
+        internal async Task<bool> IsEdgeDeviceAsync(string lnsId)
+        {
+            var keyLock = $"{nameof(EdgeDeviceGetter)}-lock";
+            var owner = nameof(EdgeDeviceGetter);
+            var isEdgeDevice = false;
+            try
+            {
+                if (await cacheStore.LockTakeAsync(keyLock, owner, TimeSpan.FromSeconds(10)))
+                {
+                    var findInCache = () => this.cacheStore.GetObject<DeviceKind>(lnsId);
+                    if (findInCache() is null)
+                    {
+                        await RefreshEdgeDevicesCacheAsync();
+                        isEdgeDevice = findInCache() is { IsEdge: true };
+                        if (!isEdgeDevice)
+                            _ = MarkDeviceAsNonEdge(lnsId);
+                    }
+                }
+                else
+                {
+                    throw new TimeoutException("Timed out while taking a lock on Redis Edge Device cache");
+                }
+            }
+            finally
+            {
+                _ = cacheStore.LockRelease(keyLock, owner);
+            }
+            return isEdgeDevice;
+        }
+
+        private bool MarkDeviceAsNonEdge(string lnsId)
+            => this.cacheStore.ObjectSet(lnsId,
```

Comment:
MarkDeviceAsNonEdge does not indicate any conditional execution. However, it is conditional and only will execute if it is not in cache. We are also not evaluating the return value when calling the method. I think the name should reflect the conditional aspect and the caller should actually emit the error if it did not update it, as that would mean, we actually had a race with the lock (potentially releasing it too early).

Reply:
Logging an error now in case of issues